### PR TITLE
[language] Add trace! to failing transactions during execution

### PR DESCRIPTION
**Pull request by dariorussi**
Add trace! for transactions that get discarded during execution.
When validating transactions, mempool already reports failures so
we don't need to do it in the VM or Adapter.
However execution was not reporting discarded transaction unless
coming from invariant violation. And discarded transactions are
not charged and a system cost.
Given that trace is not logged to the server and can be enabled
"on the fly" this is a good way to debug possible problems
without penalizing the system.

### CODE REVIEW

**Comment by dariorussi** [comment]
tested on e2e tests with logging enabled and we get all the proper tracing with good info, including `INVALID_SIGNATURE` which seems a good thing to report in execution

**Review by bmwill — APPROVED** [review]

**Review by sausagee — APPROVED** [review]

**Comment by runtian-zhou** [comment]
Are we afraid that the "false negative" is going to bother us in the future? I would imagine that those lines will be triggered when the mempool didn't reject such transaction (e.g: when the second transaction used to have enough balance to pay for gas but not until the first transaction gets committed). Is the intention to record transactions as such? Or we just want to know what are transactions that get discarded.

File: language/libra-vm/src/libra_transaction_executor.rs [hunk 1]
```diff
@@ -668,15 +676,30 @@ impl LibraVM {
...
                 }
                 Ok(PreprocessedTransaction::WriteSet(txn)) => {
-                    self.process_writeset_transaction(data_cache, *txn, &log_context)?
+                    let (vm_status, output) =
+                        self.process_writeset_transaction(data_cache, *txn, &log_context)?;
+                    (vm_status, output, Some("write_set".to_string()))
+                }
+                Err(e) => {
+                    let (vm_status, output) = discard_error_vm_status(e);
+                    (vm_status, output, None)
```
Comment (runtian-zhou) [inline language/libra-vm/src/libra_transaction_executor.rs:688]: This should be a pure signature error. Should it be `None`?

**Review by bob-wilson — APPROVED** [review]
This looks good to me

**Comment by dariorussi** [comment]
> Are we afraid that the "false negative" is going to bother us in the future? I would imagine that those lines will be triggered when the mempool didn't reject such transaction (e.g: when the second transaction used to have enough balance to pay for gas but not until the first transaction gets committed). Is the intention to record transactions as such? Or we just want to know what are transactions that get discarded.

I think we are somewhat concerned in an open system that false positive can become a problem. So in some future we may change that. But right now, and to be honest also in the future, one could argue that either those errors do not happen often (as they should not), in which case the noise is not that high anyway, and if they happen a lot being able to know about them may be a good thing. Thoughts?
In any case this is effectively logging nothing and it will only log if the level is kicked down on a given validator. And that is also the intent here, to have a debug story that is completely off in normal operation.
Let me know if you have concerns but this seems a fine story for now, particularly if we are all aware of it

File: language/libra-vm/src/libra_transaction_executor.rs (hunk 1 above)
Comment (dariorussi) [reply language/libra-vm/src/libra_transaction_executor.rs:688]: yes correct, that is what it looks like, just figured I'd bundle any error under None and print out a "malformed transaction" which for now we know to be only an invalid signature (and maybe forever?).
I somewhat did not want to be too explicit in the face of the `Err(e)`, I can change that if we think is important

**Comment by bob-wilson** [comment]
> In any case this is effectively logging nothing and it will only log if the level is kicked down on a given validator. And that is also the intent here, to have a debug story that is completely off in normal operation.
> Let me know if you have concerns but this seems a fine story for now, particularly if we are all aware of it

That matches my understanding of the situation. The last time we discussed logging these failed transactions, we decided not to do it for exactly the reason that @runtian-zhou mentioned. But, with the trace level of logging, we can selectively enable this for debugging and get that extra level of visibility only when we want it.

**Pull request closed without merging**